RUBY-1587 Properly handle error labels sent by server

saghm opened the pull request.

p-mongo commented on lib/mongo/error/operation_failure.rb:
```diff
@@ -182,6 +182,16 @@ def initialize(message = nil, result = nil, options = {})
+          if err_doc && err_doc['errorLabels']
+            err_doc['errorLabels'].each do |label|
+              add_label(label)
+            end
```
There is a Mongo::Error::Parser which parses all the other fields out of the result, how come this code is added to the exception class and not to the parser?

saghm replied:
Good idea! I'll change it to do that

p-mongo commented on spec/support/transactions/operation.rb:
```diff
@@ -114,12 +114,12 @@ def execute(collection, session0, session1)
         {
           'errorContains' => e.message,
-          'errorLabels' => e.instance_variable_get(:@labels) || []
+          'errorLabels' => e.instance_variable_get(:@labels)
```
I think Mongo::Error should have a `labels` method added following this part of https://github.com/mongodb/specifications/blob/master/source/transactions/transactions.rst#error-reporting-changes:

> Drivers MAY expose the list of all error labels for an exception object.

saghm replied:
Will do

saghm commented:
All suggested changes have been made

saghm commented:
Updated again per offline discussion

p-mongo commented on lib/mongo/operation/get_more/result.rb:
```diff
@@ -52,7 +52,7 @@ def cursor_document
 
         def first_document
-          @first_document ||= reply.documents[0]
+          @first_document ||= reply && reply.documents[0]
```
Can this be now returned to unconditionally reading `reply`?

saghm replied:
Done

saghm commented:
Sorry, I must have misunderstood how you wanted the tests changed from our offline discussion. I've added the test back and changed it to use `Result#raise_operation_failure` to ensure the correct code path is covered
